[RFC] offer: allow re-enabling a previously disabled offer

vincenzopalazzo opened the pull request:
Sometimes, for various reasons, a user disables an offer and then wants to re-enable it. This should be allowed because, from the CLN point of view, it is just an internal state.

If a user has constraints on the description of the invoice because they are using services that link some sort of user ID to an offer, it is important for the user to be able to re-enable the offer, not create a new one. Creating a new offer would require a different description.

P.S: This is in an RFC, if it will be accepted I will finish the documentation but I still need to write the test 

Fixes: https://github.com/ElementsProject/lightning/issues/7360

rustyrussell commented:
When I copied my test from the other PR, it shows why we need tests.

The final commit, which needs to be squashed, shows why...

vincenzopalazzo commented:
>The final commit, which needs to be squashed, shows why...

Yes, this RFC was not finished, I was waiting to know if we want to have an `enableoffer` in cln. Thanks for the fixup I will add the docs and push the changes by the end of the day

vincenzopalazzo commented:
>When I copied my test from the other PR, it shows why we need tests.
>The final commit, which needs to be squashed, shows why...

By accident, I force push the previous version with your code, so I remake the work, sorry about that.

NNow the code should be ready, probably I need to fix some failure with the docs, but should be a minor

Lagrang3 commented on doc/schemas/lightning-enableoffer.json:
```diff
@@ -0,0 +1,106 @@
+    ]
+  },
+  "author": [
+    "Rusty Russell <<rusty@rustcorp.com.au>> is mainly responsible."
```
Who's the author? @vincenzopalazzo or @rustyrussell

vincenzopalazzo replied:
Rusty OFC I fetch his code inside mine PR regarding the docs :)

So I forget to add the Co-Developed-by: ... thanks for chatting it

Lagrang3 commented on lightningd/offer.c:
```diff
@@ -231,6 +231,48 @@ static const struct json_command disableoffer_command = {
+		return command_fail(cmd, LIGHTNINGD, "Unknown offer");
+
+	if (offer_status_active(status))
+		return command_fail(cmd, OFFER_ALREADY_DISABLED,
```
Just a small detail: `OFFER_ALREADY_DISABLED` code does not reflect what is actually happening in this case.
Maybe the correct code should be `OFFER_ALREADY_EXISTS` or a new one `OFFER_ALREADY_ENABLED`.

vincenzopalazzo replied:
Good point thanks!

rustyrussell commented:
Rebased, and added a new return code a-la @Lagrang3 suggestion.

vincenzopalazzo commented:
Thanks for the @Lagrang3 review, sorry for the 3-day silence from my side,  but I was taking a small holiday, and now I am back.

Added Rusty ad Co-Developed-by inside the https://github.com/ElementsProject/lightning/pull/7362/commits/7bdc8db68bb9afc6d6218ec21636940b2bf20624 so when complaining will complain about this commit, can `git blame` also him :-P

In addition, I fixed a test failure regarding the code error in https://github.com/ElementsProject/lightning/pull/7362/commits/df46f0e0f3e1e39b533c4386f4aa05bc1e9d10c2 now the CI should be happy!

Thanks to everyone for keeping this PR going also when I was off <3
